File: Patterns/Abstract Factory Pattern/Cpp/Pizza/Pizza.cpp
#pragma once

#include <iostream>
#include <string>
#include <vector>

#include "./Ingredient/Dough/Dough.cpp"
#include "./Ingredient/Cheese/Cheese.cpp"
#include "./Ingredient/Sauce/Sauce.cpp"
#include "./Ingredient/Pepperoni/Pepperoni.cpp"
#include "./Ingredient/Clams/Clams.cpp"
#include "./Ingredient/Veggies/Veggies.cpp"

using namespace std;

class Pizza {
public:
    string name;

    Dough* dough;
    Sauce* sauce;
    Cheese* cheese;
    Pepperoni* pepperoni;
    Clams* clams;
    vector<Veggies*> veggies;

    virtual void prepare() = 0;

    void bake() {
        cout << "175도에서 25분 간 굽기\n";
    }

    void cut() {
        cout << "피자를 사선으로 자르기\n";
    }

    void box() {
        cout << "상자에 피자 담기\n";
    }

    void setName(string name) {
        this->name = name;
    }

    string getName() {
        return name;
    }
};

File: Patterns/Abstract Factory Pattern/CSharp/Pizza/Pizza.cs
using System;

using Ingredient = Pizza.Ingredient;

namespace Pizza
{
    public abstract class Pizza
    {
        private string name;
        public string Name { get => name; set => name = value; }

        public Ingredient.Dough.Dough? dough;
        public Ingredient.Sauce.Sauce? sauce;
        public Ingredient.Cheese.Cheese? cheese;
        public Ingredient.Pepperoni.Pepperoni? pepperoni;
        public Ingredient.Clams.Clams? clams;
        public Ingredient.Veggies.Veggies[] veggies;

        public Pizza() 
        {
            name = "일반 피자";
            dough = null;
            sauce = null;
            cheese = null;
            pepperoni = null;
            clams = null;
            veggies = new Ingredient.Veggies.Veggies[] {};
        }

        public abstract void Prepare();

        public void Bake() => Console.WriteLine("175도에서 25분 간 굽기");
        public void Cut() => Console.WriteLine("피자를 사선으로 자르기");
        public void Box() => Console.WriteLine("상자에 피자 담기");
    }
}